banner/models.py

```diff
@@ -1,189 +1,49 @@
-import re
-from random import randint
-
 from django.db import models
-from django.core.urlresolvers import reverse
+from django.utils.translation import ugettext_lazy as _
 
 from jmbo.models import ModelBase
-from preferences.models import Preferences
+from link.models import Link
+from simplemde.fields import SimpleMDEField
+
+from banner.styles import BANNER_STYLE_CLASSES
 
 
 class Banner(ModelBase):
-    """Legacy base class. Never surfaced in admin."""
-    paths = models.TextField(
-        null=True,
-        blank=True,
-        help_text="""A list of regular expressions that match the path. This \
-is used to find a banner to render when using a banner proxy. One entry per \
-line. If you are unfamiliar with regular expressions then enter the relative \
-URL to an item, eg. /post/my-post."""
+    """Base class for all banners"""
+    link = models.ForeignKey(
+        Link, help_text=_("Link to which this banner should redirect."),
+        blank=True, null=True
     )
-
-
-class CodeBanner(Banner):
-    code = models.TextField(
-        help_text='The full HTML/Javascript code snippet to be \
-embedded for this banner.'
+    style = models.CharField(
+        choices=[
+            (klass.__name__, klass.__name__)
+            for klass in BANNER_STYLE_CLASSES
+        ],
+        max_length=128, default="BaseStyle"
     )
+    headline = SimpleMDEField(null=True, blank=True, help_text=_("The banner's headline."))
+    body = SimpleMDEField(null=True, blank=True, help_text=_("The banner's main text content."))
 
 
-class BaseImageBanner(Banner):
-    """For every different aspect ratio of image banner a subclass is required.
-    Use this class as base."""
-    url = models.CharField(
-        max_length='256',
-        verbose_name='URL',
-        help_text='URL (internal or external) to which this banner will link.'
+class Button(models.Model):
+    """Call to action handling"""
+    text = models.CharField(
+        max_length=60,
+        help_text=_("The text to be displayed as the button label")
     )
-    dfp_slot_name = models.CharField(
-        max_length=256,
-        null=True,
-        blank=True,
-        help_text="""A combination of network code and ad unit as provided \
-by Google, eg. /1234/travel. Used to track clicks."""
+    link = models.ForeignKey(
+        Link, help_text=_("CTA link for this button"), null=True, blank=True
     )
-    dfp_ad_id = models.CharField(
-        max_length=64,
-        null=True,
-        blank=True,
-        help_text="The advertisement id. Used to track clicks."
+    banner = models.ManyToManyField(
```

Comment:
Consider using sortedm2m field for this. It works around some Django query bugs for this type of use case.

Reply:
:+1: